uniffi/src/support/tests.rs

```diff
@@ -0,0 +1,141 @@
+/// component is compiled and available for use and then executes the foreign language script,
+/// returning successfully iff the script exits successfully.
+pub fn run_foreign_language_testcase<P: AsRef<Path>>(pkg_dir: &str, test_file: P) -> Result<()> {
+    let test_file: &Path = test_file.as_ref();
+    let cdylib_file = ensure_compiled_cdylib(pkg_dir)?;
+    let out_dir = Path::new(cdylib_file.as_str())
+        .parent()
+        .ok_or_else(|| anyhow::anyhow!("Generated cdylib has not parent directory"))?;
+    let lang = test_file.extension().unwrap_or_default();
+    ensure_generated_bindings(cdylib_file.as_str(), out_dir, lang)?;
+    bindings::run_script(Some(out_dir), Some(test_file), lang.try_into()?)
+}
+
+/// Ensure that a uniffi component crate is compiled and ready for use.
+///
+/// This function takes the top-level directory of a uniffi component crate, ensures that the
+/// component's cdylib is compiled and available for use in generating bindings and running
+/// foreign language code.
+///
+/// Internally, this function does a bit of caching and concurrency management to avoid rebuilding
+/// the component for multiple testcases.
+pub fn ensure_compiled_cdylib(pkg_dir: &str) -> Result<String> {
+    // Have we already compiled this component?
+    let mut compiled_components = COMPILED_COMPONENTS.lock().unwrap();
+    if let Some(cdylib_file) = compiled_components.get(pkg_dir) {
+        return Ok(cdylib_file.to_string());
+    }
+    // Nope, looks like we'll have to compile it afresh.
+    let mut cmd = Command::new("cargo");
+    cmd.arg("build").arg("--message-format=json").arg("--lib");
+    cmd.current_dir(pkg_dir);
+    cmd.stdout(Stdio::piped());
+    let mut child = cmd.spawn()?;
+    let output = std::io::BufReader::new(child.stdout.take().unwrap());
+    // Build the crate, looking for any cdylibs that it might produce.
+    let cdylibs = cargo_metadata::Message::parse_stream(output)
+        .filter_map(|message| match message {
+            Err(e) => Some(Err(e.into())),
+            Ok(Message::CompilerArtifact(artifact)) => {
+                if artifact.target.kind.iter().any(|item| item == "cdylib") {
+                    Some(Ok(artifact))
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        })
+        .collect::<Result<Vec<_>>>()?;
+    if !child.wait()?.success() {
+        bail!("Failed to execute `cargo build`");
+    }
+    // If we didn't just build exactly one cdylib, we're going to have a bad time.
+    match cdylibs.len() {
+        0 => bail!("Crate did not produce any cdylibs, it must not be a uniffi component"),
+        1 => (),
+        _ => bail!("Crate produced multiple cdylibs, it must not be a uniffi component"),
+    }
+    let cdylib_files: Vec<_> = cdylibs[0]
+        .filenames
+        .iter()
+        .filter(|nm| match nm.extension().unwrap_or_default().to_str() {
+            Some("dylib") | Some("so") => true,
```

Comment:
Would this be `dll` for Windows someday?

Reply:
Maybe one day, when @mhammond gets sufficiently ~~annoyed~~ motivated to shave that yak 😁